src/librustc_typeck/check/pat.rs

```diff
@@ -143,13 +146,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 self.check_pat_tuple_struct(pat, qpath, subpats, ddpos, expected, def_bm, ti)
             }
             PatKind::Path(ref qpath) => {
-                self.check_pat_path(pat, path_resolution.unwrap(), qpath, expected)
+                self.check_pat_path(pat, path_resolution.unwrap(), qpath, expected, ti.parent_pat)
             }
             PatKind::Struct(ref qpath, fields, etc) => {
                 self.check_pat_struct(pat, qpath, fields, etc, expected, def_bm, ti)
             }
             PatKind::Or(pats) => {
+                let parent_pat = &pat;
                 for pat in pats {
+                    let mut ti = ti;
+                    ti.parent_pat = Some(parent_pat);
```

Comment:
Nit / general note: I think it would be slightly more readable to use FRU instead, `TopInfo { parent_pat: Some(&pat), ..ti }` (with a binding, or not, if it fits in the same line as the `.check_pat`).